src/include/dng/hts/bcf.h

```
@@ -0,0 +1,293 @@
+
+#include "hts.h"
+
+#include <vector>
+#include <string>
+#include <set>
+#include <htslib/vcf.h>
+
+
+#define VCF_VERSION "VCFv4.1"
+
+namespace hts { namespace bcf { 
+
+/**
+ * File - Class for writing VCF/BCF files to the output, either stdout or to a file
+ * To write to stdout set file="-" in constructor. To write as a BCF file set mode="wb" in constructor.
+ * To use:
+ *  1. call addHeaderMetadata() and addSample() to add information to the VCF header
+ *  2. call WriteHdr() to finish the header
+ *  3. populate the first few columns of a VCF body line using SetID(), SetFilter(), SetAlleles(), UpdateInfo() and UpdateSamples() 
+ *  4. call WriteRecord() to save/print the current VCF line and move on to the next
+ */
+class File : hts::File
+{
+ public:
+  
+  /**
+   * Initialize File/output stream for writing VCF/BCF
+   * @file - file name to write to, use "-" for stdout
+   * @mode: "w" = write VCF, "wb" = write BCF
+   * @source: name of application writing the VCF file
+   */
+  File(const char *file, const char *mode, const char *source)
+    : hts::File(file, mode)
+    {
+      hdr = bcf_hdr_init("w"); // VCF header
+      rec = bcf_init1(); // VCF record/body line
+      
+      std::string s = std::string("#source=") + source;
+      bcf_hdr_append(hdr, s.c_str());
+    }
+
+  File(const File&) = delete;
+
+  File& operator=(File& other)
+    {
+      if(this == &other)
+	return *this;
+      
+      hdr = other.hdr;
+      other.hdr = nullptr;
+
+      rec = other.rec;
+      other.rec = nullptr;
+      
+      contig_ids = other.contig_ids;
+      other.contig_ids.clear();
+
+      return *this;
+    }
+      
+  virtual ~File()
+    {
+      if(rec != nullptr)
+	bcf_destroy1(rec);
+
+      if(hdr != nullptr)
+	bcf_hdr_destroy(hdr);
+
+    }
+
+  /** AddHeaderMetadata() - Adds a "##key=value" line to the VCF header */
+  void AddHeaderMetadata(const char *key, const char *value)
+  {
+    if(key == nullptr || value == nullptr)
```

What does bcf_hrd_append do if these are null?  Is it important to silently ignore this?